File: _legacy/backend_python_v1/tools/screen_capture.py
import base64
import io
import os
import sys
from typing import Dict, Any

import mss
from PIL import Image

try:
    from ..core.base_tool import BaseTool
except ImportError:
    # Fallback when imported as a top-level module (no parent package)
    if __package__ and __package__.startswith("backend"):
        raise
    backend_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    if backend_dir not in sys.path:
        sys.path.insert(0, backend_dir)
    from core.base_tool import BaseTool

class ScreenCaptureTool(BaseTool):
    name = "screen_capture"
    description = "Capture a screenshot of the primary monitor. Returns a base64 encoded image."
    input_schema = {
        "type": "object",
        "properties": {
            "resize_factor": {
                "type": "number",
                "description": "Factor to resize the screenshot by (0.0 to 1.0). Default is 0.5.",
                "default": 0.5
            }
        },
        "required": []
    }

    async def execute(self, resize_factor: float = 0.5) -> Dict[str, Any]:
        with mss.mss() as sct:
            # Capture the primary monitor
            monitor = sct.monitors[1]
            sct_img = sct.grab(monitor)

            # Convert to PIL Image
            img = Image.frombytes("RGB", sct_img.size, sct_img.bgra, "raw", "BGRX")

            # Resize if needed
            if resize_factor != 1.0:
                new_size = (int(img.width * resize_factor), int(img.height * resize_factor))
                img = img.resize(new_size, Image.Resampling.LANCZOS)

            # Convert to base64
            buffered = io.BytesIO()
            img.save(buffered, format="JPEG", quality=85)
            img_str = base64.b64encode(buffered.getvalue()).decode()

            return {
                "type": "image",
                "data": f"data:image/jpeg;base64,{img_str}",
                "width": img.width,
                "height": img.height
            }

File: backend-node/src/tools/screen_capture.ts
const screenshot = require('screenshot-desktop');
import sharp from 'sharp';
import { BaseToolImplementation } from '../core/base_tool';

export class ScreenCaptureTool extends BaseToolImplementation {
    name = "screen_capture";
    description = "Capture a screenshot of the primary monitor. Returns a base64 encoded image.";
    input_schema = {
        type: "object",
        properties: {
            resize_factor: {
                type: "number",
                description: "Factor to resize the screenshot by (0.0 to 1.0). Default is 0.5.",
                default: 0.5
            }
        },
        required: []
    };

    async execute(args: { resize_factor?: number }, _context: { sessionId?: string } = {}): Promise<any> {
        const resizeFactor = args.resize_factor ?? 0.5;
        
        try {
            const imgBuffer = await screenshot({ format: 'png' });
            
            let finalBuffer = imgBuffer;
            let width = 0;
            let height = 0;

            const metadata = await sharp(imgBuffer).metadata();
            width = metadata.width || 0;
            height = metadata.height || 0;

            if (resizeFactor !== 1.0) {
                const newWidth = Math.round(width * resizeFactor);
                finalBuffer = await sharp(imgBuffer)
                    .resize({ width: newWidth })
                    .jpeg({ quality: 85 })
                    .toBuffer();
            } else {
                 finalBuffer = await sharp(imgBuffer)
                    .jpeg({ quality: 85 })
                    .toBuffer();
            }
            
            // Re-read metadata after resize/convert
            const newMeta = await sharp(finalBuffer).metadata();
            width = newMeta.width || 0;
            height = newMeta.height || 0;

            // Convert to base64
            const base64Image = finalBuffer.toString('base64');
            
            return {
                type: "image",
                data: `data:image/jpeg;base64,${base64Image}`,
                width: width,
                height: height
            };

        } catch (e: any) {
            return { status: "error", message: e.message };
        }
    }
}
